### PR TITLE
Introduce 'transfers' packages

### PR DESCRIPTION
**Pull request by potiuk**
---
Make sure to mark the boxes below before creating PR: [x]

- [x] Description above provides context of the change
- [x] Unit tests coverage for changes (not needed for documentation changes)
- [x] Target Github ISSUE in description if exists
- [x] Commits follow "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)"
- [x] Relevant documentation is updated including usage instructions.
- [x] I will engage committers as explained in [Contribution Workflow Example](https://github.com/apache/airflow/blob/master/CONTRIBUTING.rst#contribution-workflow-example).

---
In case of fundamental code change, Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvements+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in [UPDATING.md](https://github.com/apache/airflow/blob/master/UPDATING.md).
Read the [Pull Request Guidelines](https://github.com/apache/airflow/blob/master/CONTRIBUTING.rst#pull-request-guidelines) for more information.

### CODE REVIEW

File: UPDATING.md [hunk 1]
```diff
@@ -971,8 +971,8 @@ The following table shows changes in import paths.
...
 |airflow.contrib.operators.gcp_spanner_operator.CloudSpannerInstanceDeployOperator                                 |airflow.providers.google.cloud.operators.spanner.SpannerDeployInstanceOperator                                                |
-|airflow.contrib.operators.gcp_speech_to_text_operator.GcpSpeechToTextRecognizeSpeechOperator                      |airflow.providers.google.cloud.operators.speech_to_text.CloudSpeechToTextRecognizeSpeechOperator                              |
-|airflow.contrib.operators.gcp_text_to_speech_operator.GcpTextToSpeechSynthesizeOperator                           |airflow.providers.google.cloud.operators.text_to_speech.CloudTextToSpeechSynthesizeOperator                                   |
+|airflow.contrib.operators.gcp_speech_to_text_operator.GcpSpeechToTextRecognizeSpeechOperator                      |airflow.providers.google.cloud.transfers.speech_to_text.CloudSpeechToTextRecognizeSpeechOperator                              |
```
Comment (feluelle) [inline UPDATING.md]: That’s an exception. This should not be under transfer?! Transfer Operators are services which move data from one into another, correct?

File: UPDATING.md [hunk 2]
```diff
@@ -971,8 +971,8 @@ The following table shows changes in import paths.
...
+|airflow.contrib.operators.gcp_text_to_speech_operator.GcpTextToSpeechSynthesizeOperator                           |airflow.providers.google.cloud.transfers.text_to_speech.CloudTextToSpeechSynthesizeOperator                                   |
```
Comment (feluelle) [inline UPDATING.md]: Same here.

Comment (potiuk) [reply UPDATING.md]: Ah yeah.

File: CONTRIBUTING.rst [hunk 3]
```diff
@@ -542,8 +542,48 @@ We support the following types of tests:
...
+    * hooks -> hooks are stored here
+    * operators -> operators are stored here
+    * sensors -> sensors are stored here
+    * secrets -> secret backends are stored here
+    * transfers -> transfer operators are stored here
+* Module names do not contain word "hooks" , "operators" etc. The right type comes from
+  the package. For example 'hooks.datastore' module contains DataStore hook and 'operators.datastore'
+  contains DataStore operators.
+* Class names contain 'Operator', 'Hook', 'Sensor' - for example DataStoreHook, DataStoreExportOperator
+* Operator name usually follows the convention: <Subject><Action>Operator (DataStoreExportOperator) is a good
```
Comment (mik-laj) [inline CONTRIBUTING.rst]:
```suggestion
* Operator name usually follows the convention: <Subject><Action><Entity>Operator (BigQueryExecuteQueryOperator) is a good
```
It is not a good idea to use the verb alone. Here is an example: https://github.com/apache/airflow/pull/9314
We had CloudBuildCreate to create Build, and then it turned out that the service has grown and you can also create BuildTriggey, so we created a new operator CloudBuildCreateBuildTrigger. If we do not have a noun, the operator's name has become ambiguous.

**Comment by mik-laj** [comment]
Can you view the file operators-and-hooks-ref.rst? It has sections with transfer operators for a long time, so it's easy to find a bug there.
Example:
:mod:`airflow.providers.amazon.aws.operators.sftp_to_s3` in Transfer operators and hooks

I think we should also add transfer package to check
https://github.com/PolideaInternal/airflow/blob/backport-packages-rename/docs/build#L171

File: UPDATING.md (hunk 1 above)
Comment (potiuk) [reply UPDATING.md]: Yep. REsolved

**Comment by potiuk** [comment]
> Can you view the file operators-and-hooks-ref.rst? It has sections with transfer operators for a long time, so it's easy to find a bug there.
> Example:
> :mod:`airflow.providers.amazon.aws.operators.sftp_to_s3` in Transfer operators and hooks
> 
> I think we should also add transfer package to check
> https://github.com/PolideaInternal/airflow/blob/backport-packages-rename/docs/build#L171

@mik-laj -> I think all the concerns are addressed now. I also reviewed operators-and-hooks.rst and it looks all should be fine now.

File: airflow/operators/google_api_to_s3_transfer.py [hunk 4]
```diff
@@ -17,32 +17,32 @@
...
     This class is deprecated.
     Please use:
-    `airflow.providers.amazon.aws.operators.google_api_to_s3_transfer.GoogleApiToS3TransferOperator`."""
+    `airflow.providers.amazon.aws.transfers.google_api_to_s3_transfer.GoogleApiToS3Operator`."""
 
     def __init__(self, *args, **kwargs):
         warnings.warn(
             """This class is deprecated.
             Please use
             `airflow.providers.amazon.aws.operators.""" +
```
Comment (mik-laj) [inline airflow/operators/google_api_to_s3_transfer.py]:
```suggestion
            `airflow.providers.amazon.aws.transfers.""" +
```

File: airflow/providers/google/README.md [hunk 5]
```diff
@@ -421,7 +406,40 @@ All classes in Airflow 2.0 are in `airflow.providers.google` package.
...
+| [cloud.transfers.gcs_to_local.GCSToLocalOperator](https://github.com/apache/airflow/blob/master/airflow/providers/google/cloud/transfers/gcs_to_local.py)                           |
+| [cloud.transfers.gcs_to_sftp.GCSToSFTPOperator](https://github.com/apache/airflow/blob/master/airflow/providers/google/cloud/transfers/gcs_to_sftp.py)                              |
+| [cloud.transfers.presto_to_gcs.PrestoToGCSOperator](https://github.com/apache/airflow/blob/master/airflow/providers/google/cloud/transfers/presto_to_gcs.py)                        |
+| [cloud.transfers.sftp_to_gcs.SFTPToGCSOperator](https://github.com/apache/airflow/blob/master/airflow/providers/google/cloud/transfers/sftp_to_gcs.py)                              |
+| [cloud.transfers.sheets_to_gcs.GoogleSheetsToGCSOperator](https://github.com/apache/airflow/blob/master/airflow/providers/google/cloud/transfers/sheets_to_gcs.py)                  |
+| [suite.transfers.gcs_to_sheets.GCSToGoogleSheetsOperator](https://github.com/apache/airflow/blob/master/airflow/providers/google/suite/transfers/gcs_to_sheets.py)                  |
+
+
+
+    ### Moved transfer operators
```
Comment (mik-laj) [inline airflow/providers/google/README.md]: Is this indentation expected?

File: docs/operators-and-hooks-ref.rst [hunk 6]
```diff
@@ -530,63 +530,63 @@ These integrations allow you to copy data from/to Amazon Web Services.
...
-     - :mod:`airflow.providers.apache.hive.operators.s3_to_hive`
+     - :mod:`airflow.providers.apache.hive.transfers.s3_to_hive`
 
    * - `Amazon Simple Storage Service (S3) <https://aws.amazon.com/s3/>`__
      - `Google Cloud Storage (GCS) <https://cloud.google.com/gcs/>`__
      - :doc:`How to use <howto/operator/gcp/cloud_storage_transfer_service>`
-     - :mod:`airflow.providers.google.cloud.operators.s3_to_gcs`,
-       :mod:`airflow.providers.google.cloud.operators.cloud_storage_transfer_service`
+     - :mod:`airflow.providers.google.cloud.transfers.s3_to_gcs`,
+       :mod:`airflow.providers.google.cloud.transfers.cloud_storage_transfer_service`
```
Comment (mik-laj) [inline docs/operators-and-hooks-ref.rst]: Cna you check it? Is it ok?

**Review by mik-laj — APPROVED** [review]

File: airflow/providers/google/README.md (hunk 5 above)
Comment (potiuk) [reply airflow/providers/google/README.md]: Nope. Fixed.

File: docs/operators-and-hooks-ref.rst (hunk 6 above)
Comment (potiuk) [reply docs/operators-and-hooks-ref.rst]: What do you mean by OK ? It looks good. What do you suspect is wrong here?

**Comment by potiuk** [comment]
@mik-laj -> all corrected, just the last comment I am not sure about

**Comment by potiuk** [comment]
@feluelle  - one more chance to take a look before I merge :)

File: docs/operators-and-hooks-ref.rst (hunk 6 above)
Comment (mik-laj) [reply docs/operators-and-hooks-ref.rst]: Cloud transfer service operators should be in operators package.

**Comment by potiuk** [comment]
Added more descriptions for Transfer packages in CONTRIBUTING.rst and added transfer packages to be checked for consistency (added two missing modules and one missing howto.

**Review by mik-laj — APPROVED** [review]

File: airflow/providers/mysql/transfers/s3_to_mysql.py [hunk 7]
```diff
@@ -24,7 +24,7 @@
 from airflow.utils.decorators import apply_defaults
 
 
-class S3ToMySqlTransferOperator(BaseOperator):
+class S3ToMySqlOperator(BaseOperator):
     """
     Loads a file from S3 into a MySQL table.
```
Comment (jensenity) [inline airflow/providers/mysql/transfers/s3_to_mysql.py:29]: Can I ask what's the requirement S3 file format for this Operator? json? csv? parquet and etc.

Comment (feluelle) [reply airflow/providers/mysql/transfers/s3_to_mysql.py:29]: It calls [bulk_load_custom](https://github.com/apache/airflow/blob/68d1714f296989b7aad1a04b75dc033e76afb747/airflow/providers/mysql/hooks/mysql.py#L217) which uses [LOAD DATA](https://dev.mysql.com/doc/refman/8.0/en/load-data.html)

Comment (feluelle) [reply airflow/providers/mysql/transfers/s3_to_mysql.py:29]: > The LOAD DATA statement reads rows from a text file into a table at a very high speed.

You can define sth like:
```
    [{FIELDS | COLUMNS}
        [TERMINATED BY 'string']
        [[OPTIONALLY] ENCLOSED BY 'char']
        [ESCAPED BY 'char']
    ]
    [LINES
        [STARTING BY 'string']
        [TERMINATED BY 'string']
    ]
```
for csv files.

Comment (jensenity) [reply airflow/providers/mysql/transfers/s3_to_mysql.py:29]: Thanks @feluelle, this helped a lot!

Comment (jensenity) [reply airflow/providers/mysql/transfers/s3_to_mysql.py:29]: One more question! Does this work with parquet files? After looking into the docs, I don't think it does. Right?

Comment (feluelle) [reply airflow/providers/mysql/transfers/s3_to_mysql.py:29]: No I don't think it works directly. You would have to convert parquet to csv and load csv to mysql.

**Pull request merged**